velox/expression/CastExpr.cpp

```diff
@@ -146,6 +146,107 @@ void propagateErrorsOrSetNulls(
 }
 } // namespace
 
+#define VELOX_DYNAMIC_DECIMAL_TYPE_DISPATCH(       \
+    TEMPLATE_FUNC, decimalTypePtr, ...)            \
+  [&]() {                                          \
+    if (decimalTypePtr->isLongDecimal()) {         \
+      return TEMPLATE_FUNC<int128_t>(__VA_ARGS__); \
+    } else {                                       \
+      return TEMPLATE_FUNC<int64_t>(__VA_ARGS__);  \
+    }                                              \
+  }()
+
+template <typename FromNativeType, TypeKind ToKind>
+VectorPtr applyDecimalToIntegralCast(
+    const SelectivityVector& rows,
+    const BaseVector& input,
+    exec::EvalCtx& context,
+    const TypePtr& fromType,
+    const TypePtr& toType) {
+  using To = typename TypeTraits<ToKind>::NativeType;
+
+  VectorPtr result;
+  context.ensureWritable(rows, toType, result);
+  (*result).clearNulls(rows);
+  auto resultBuffer = result->asUnchecked<FlatVector<To>>()->mutableRawValues();
+  const auto precisionScale = getDecimalPrecisionScale(*fromType);
+  const auto simpleInput = input.as<SimpleVector<FromNativeType>>();
+  const auto scaleFactor = DecimalUtil::kPowersOfTen[precisionScale.second];
+  context.applyToSelectedNoThrow(rows, [&](int row) {
+    auto value = simpleInput->valueAt(row);
+    auto integralPart = value / scaleFactor;
+    auto fractionPart = value % scaleFactor;
+    auto sign = value >= 0 ? 1 : -1;
+    bool needsRoundUp =
+        (scaleFactor != 1) && (sign * fractionPart >= (scaleFactor >> 1));
+    integralPart += needsRoundUp ? sign : 0;
+    if (integralPart > std::numeric_limits<To>::max() ||
+        integralPart < std::numeric_limits<To>::min())
+      VELOX_USER_FAIL(makeErrorMessage(input, row, toType) + "Out of bounds.");
+
+    resultBuffer[row] = static_cast<To>(integralPart);
+  });
+  return result;
+}
+
+template <typename FromNativeType, TypeKind ToKind>
+VectorPtr applyDecimalToFloatCast(
+    const SelectivityVector& rows,
+    const BaseVector& input,
+    exec::EvalCtx& context,
+    const TypePtr& fromType,
+    const TypePtr& toType) {
+  using To = typename TypeTraits<ToKind>::NativeType;
+
+  VectorPtr result;
+  context.ensureWritable(rows, toType, result);
+  (*result).clearNulls(rows);
+  auto resultBuffer = result->asUnchecked<FlatVector<To>>()->mutableRawValues();
+  const auto precisionScale = getDecimalPrecisionScale(*fromType);
+  const auto simpleInput = input.as<SimpleVector<FromNativeType>>();
+  const auto scaleFactor = DecimalUtil::kPowersOfTen[precisionScale.second];
+  context.applyToSelectedNoThrow(rows, [&](int row) {
+    auto output =
+        util::Converter<ToKind, void, false>::cast(simpleInput->valueAt(row));
```

Comment:
We are seeing poor performance of try_cast when many values throw. @laithsakka optimized some of the code paths to avoid throwing. Would it be possible to optimize this code path as well? See https://github.com/facebookincubator/velox/pull/5801

Reply:
Done.